service/s3/s3crypto/crypto.go

```diff
@@ -0,0 +1,22 @@
+package s3crypto
+
+import (
+	"bytes"
+	"io"
+)
+
+// Cipher placeholder
+type Cipher interface {
+	Encrypter
+	Decrypter
+}
+
+// Encrypter interface with only the encrypt method
+type Encrypter interface {
```

Comment:
Rather than attempting to adapt all the ciphers behind a stream-like API, what do you think about using the core golang crypto [cipher.Block](https://golang.org/pkg/crypto/cipher/#Block) and [cipher.Stream](https://golang.org/pkg/crypto/cipher/#Stream) interfaces and using type assertions instead? As it stands, all of the algorithms currently implemented call `ioutil.ReadAll` immediately rather than use streaming crypto algorithms. Even the  CTR methods are not using [cipher.NewCTR](https://golang.org/pkg/crypto/cipher/#NewCTR), which could simplify the implementation significantly.